### PR TITLE
[REF] web: adds the callback onWillSaveRecord and onRecordSaved

### PR DESCRIPTION
**Pull request by FrancoisGe**
The purpose of this commit is to add extension points that allow code
to be executed before and after the save of a record.

CallBack:
    onWillSaveRecord is a callBack that will be executed before the
        record save. If it returns false, it will prevent the save.

    onSavedRecord is a callBack that will be executed after the save
        if it was done.
This commit will replace all the save overwrites by the onWillSaveRecord
and onSavedRecord callBack.

Observed problem:
We could notice that each of the overwrites of save in order to execute
for example a doAction after this one did not take into account the fact
that the save did not succeed.
For example, a required field is invalid and we try to save, then the
doAction will be executed without the "write" of the save.
The callBack onSavedRecord will prevent this common error.
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr/static/src/views/profile_form_view.js [hunk 1]
```diff
@@ -1,25 +1,25 @@
...
+    }
+
+    async afterSaveRecord(record) {
+        const dirtyFields = record.dirtyFields.map((f) => f.name);
         if (dirtyFields.includes("lang")) {
-            this.model.actionService.doAction("reload_context");
+            await this.action.doAction("reload_context");
         }
-        return res;
+        return super.afterSaveRecord(record);
```
Comment (aab-odoo) [inline addons/hr/static/src/views/profile_form_view.js:18]: not necessary

File: addons/web/static/src/views/form/form_controller.js [hunk 2]
```diff
@@ -255,6 +257,10 @@ export class FormController extends Component {
         }
     }
 
+    async onSavedRecord(record) {}
+
+    async onWillSaveRecord(record) {}
```
Comment (aab-odoo) [inline addons/web/static/src/views/form/form_controller.js:262]: these two could be documented (explain why they exist, when they are executed and what should be their API if they are overriden)

File: addons/web/static/src/views/list/list_controller.js [hunk 3]
```diff
@@ -155,6 +157,10 @@ export class ListController extends Component {
         );
     }
 
+    async onSavedRecord() {}
+
+    async onWillSaveRecord() {}
+
```
Comment (aab-odoo) [inline addons/web/static/src/views/list/list_controller.js:163]: same

File: addons/web/static/src/views/relational_model.js [hunk 4]
```diff
@@ -435,6 +435,9 @@ export class Record extends DataPoint {
         this._onWillSwitchMode = params.onRecordWillSwitchMode || (() => {});
 
         this.canBeAbandoned = this.isVirtual;
+
+        this.onWillSaveRecord = params.onWillSaveRecord || (() => {});
+        this.onSavedRecord = params.onSavedRecord || (() => {});
```
Comment (aab-odoo) [inline addons/web/static/src/views/relational_model.js:440]: can't this be done once, in Datapoint?

File: addons/web/static/src/views/relational_model.js [hunk 5]
```diff
@@ -1813,6 +1823,22 @@ export class DynamicRecordList extends DynamicList {
...
+            activeFields: this.activeFields,
+            onRecordWillSwitchMode: this.onRecordWillSwitchMode,
+            defaultContext: this.defaultContext,
+            rawContext: {
+                parent: this.rawContext,
+                make: () => this.context,
+            },
+            onWillSaveRecord: this.onWillSaveRecord,
+            onSavedRecord: this.onSavedRecord,
+        };
```
Comment (aab-odoo) [inline addons/web/static/src/views/relational_model.js:1839]: maybe sort this somehow (at least regroup callbacks)

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 6]
```diff
@@ -12573,4 +12573,91 @@ QUnit.module("Views", (hooks) => {
             "Unable to save"
         );
     });
+
+    QUnit.test(
+        "execute an action before and after each valid save in a form view",
+        async function (assert) {
+            const formView = registry.category("views").get("form");
+            class FormViewCustom extends formView.Controller {
```
Comment (aab-odoo) [inline addons/web/static/tests/views/form/form_view_tests.js:12581]: no big deal, but FormViewCustom is franglish (CustomFormView is better)

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 7]
```diff
@@ -12573,4 +12573,91 @@ QUnit.module("Views", (hooks) => {
...
+                    assert.step(`onWillSaveRecord ${record.resId}`);
+                }
+            }
+            registry.category("views").add(
+                "form",
+                {
+                    ...formView,
+                    Controller: FormViewCustom,
+                },
+                { force: true }
```
Comment (aab-odoo) [inline addons/web/static/tests/views/form/form_view_tests.js:12596]: I'd use a js_class instead

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 8]
```diff
@@ -12573,4 +12573,91 @@ QUnit.module("Views", (hooks) => {
...
+                mockRPC(route, args) {
+                    if (args.method === "write") {
+                        assert.step(`write ${args.args[0]}`);
+                    }
+                },
+            });
+
+            await editInput(target, "[name='foo'] input", "");
+            await clickSave(target);
+            assert.verifySteps(["onWillSaveRecord 1"]);
```
Comment (aab-odoo) [inline addons/web/static/tests/views/form/form_view_tests.js:12614]: this highlights a problem imo. This hook should only be called if the record will really be saved, don't you think?

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 9]
```diff
@@ -12573,4 +12573,91 @@ QUnit.module("Views", (hooks) => {
...
+
+            await editInput(target, "[name='foo'] input", "YOLO");
+            await clickSave(target);
+            assert.verifySteps(["onWillSaveRecord 1", "write 1", "onSavedRecord 1"]);
+        }
+    );
+
+    QUnit.test(
+        "execute an action before and after each valid save in a form view",
```
Comment (aab-odoo) [inline addons/web/static/tests/views/form/form_view_tests.js:12623]: same name as previous test

File: addons/web/static/tests/views/list_view_tests.js [hunk 10]
```diff
@@ -1971,6 +1971,139 @@ QUnit.module("Views", (hooks) => {
...
+        }
+    );
+
+    QUnit.test(
+        "don't exec a valid save with onWillSaveRecord in a list view",
+        async function (assert) {
+            const listView = registry.category("views").get("list");
+            class ListViewCustom extends listView.Controller {
+                async onSavedRecord(record) {
+                    assert.step(`onSavedRecord ${record.resId}`);
```
Comment (aab-odoo) [inline addons/web/static/tests/views/list_view_tests.js:2071]: I'd throw an error, this is more explicit

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 11]
```diff
@@ -12573,4 +12573,91 @@ QUnit.module("Views", (hooks) => {
...
+        async function (assert) {
+            const formView = registry.category("views").get("form");
+            class FormViewCustom extends formView.Controller {
+                async onSavedRecord(record) {
+                    assert.step(`onSavedRecord ${record.resId}`);
```
Comment (aab-odoo) [inline addons/web/static/tests/views/form/form_view_tests.js:12628]: I'd throw an Error, this is more explicit

File: addons/crm/static/src/views/crm_form/crm_form.js [hunk 12]
```diff
@@ -58,34 +66,26 @@ export class CrmFormRecord extends Record {
...
         }
-        return res;
     }
-}
 
-class CrmFormModel extends RelationalModel {
-    setup(params, services) {
-        this.effect = services.effect;
-        super.setup(...arguments);
+    async onSavedRecord(record) {
```
Comment (aab-odoo) [inline addons/crm/static/src/views/crm_form/crm_form.js:81]: onRecordSaved sounds better imo

**Review by aab-odoo — COMMENTED** [review]
Interesting work :-)

File: addons/web/static/tests/views/form/form_view_tests.js (hunk 8 above)
Comment (FrancoisGe) [reply addons/web/static/tests/views/form/form_view_tests.js:12614]: You wan't to do it after the checkValidity ?

File: addons/web/static/tests/views/list_view_tests.js (hunk 10 above)
Comment (FrancoisGe) [reply addons/web/static/tests/views/list_view_tests.js]: Why ?

File: addons/web/static/src/views/relational_model.js (hunk 5 above)
Comment (FrancoisGe) [reply addons/web/static/src/views/relational_model.js:1839]: I don't understand ?

**Review by aab-odoo — APPROVED** [review]
:+1: 

robodoo r+

**Review by aab-odoo — APPROVED** [review]
robodoo r+

**Pull request closed without merging**